# subprojects/packagefiles/syntax-transform/3.1/0038-basic-object.patch
#
# This file is part of mkxp.
#
# Copyright (C) 2026 The mkxp-z authors
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# The class `BasicObject` doesn't exist until Ruby 1.9. This patch removes all
# traces of its existence if syntax transform is targeting an earlier Ruby
# version.

--- a/eval.c
+++ b/eval.c
@@ -443,6 +443,15 @@ rb_mod_s_constants(int argc, VALUE *argv, VALUE mod)
             }
         }
 
+        /* If syntax transform is active and targeting Ruby <= 1.8, don't include `BasicObject` in the output of `Module::constants` */
+        if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+            ID id = rb_intern("BasicObject");
+            VALUE symbol = RB_ID2SYM(id);
+            if (rb_ary_includes(ary, symbol) && (!rb_const_defined(cbase, id) || rb_const_get(cbase, id) == rb_cBasicObject)) {
+                rb_ary_delete(ary, symbol);
+            }
+        }
+
         /* If syntax transform is active and targeting Ruby <= 1.8, convert the output of `Module::constants` from an array of symbols to an array of strings */
         if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
             long length = RARRAY_LEN(ary);
--- a/variable.c
+++ b/variable.c
@@ -2671,6 +2671,10 @@ static VALUE
 rb_const_get_0(VALUE klass, ID id, int exclude, int recurse, int visibility)
 {
     VALUE c = rb_const_search(klass, id, exclude, recurse, visibility);
+    /* If syntax transform is active and targeting Ruby <= 1.8, don't allow retrieving `BasicObject` */
+    if (mkxp_ec_is_syntax_transform_active(1, 8, -1) && c == rb_cBasicObject) {
+        return rb_const_missing(klass, ID2SYM(id));
+    }
     if (c != Qundef) {
         if (UNLIKELY(!rb_ractor_main_p())) {
             if (!rb_ractor_shareable_p(c)) {
@@ -3091,6 +3095,11 @@ rb_const_defined_0(VALUE klass, ID id, int exclude, int recurse, int visibility)
 		return (int)Qfalse;
 	    }
 
+	    /* If syntax transform is active and targeting Ruby <= 1.8, make sure `BasicObject` isn't defined */
+	    if (mkxp_ec_is_syntax_transform_active(1, 8, -1) && ce->value == rb_cBasicObject) {
+		return (int)Qfalse;
+	    }
+
 	    return (int)Qtrue;
 	}
 	if (!recurse) break;
